internal/mode/static/state/graph/policies.go

```diff
@@ -0,0 +1,313 @@
+	Conditions []conditions.Condition
+}
+
+// PolicyAncestorRef contains the identifying information of the ancestor.
+type PolicyAncestorRef struct {
+	// Kind is the Kind of the object.
+	Kind v1.Kind
+	// Group is the Group of the object.
+	Group v1.Group
+	// Nsname is the NamespacedName of the object.
+	Nsname types.NamespacedName
+	// SectionName is the SectionName of the object. For example, a listener name. This may be empty.
+	SectionName string
+}
+
+// PolicyTargetRef represents the object that the Policy is targeting.
+type PolicyTargetRef struct {
+	// Kind is the Kind of the object.
+	Kind v1.Kind
+	// Group is the Group of the object.
+	Group v1.Group
+	// Nsname is the NamespacedName of the object.
+	Nsname types.NamespacedName
+}
+
+// PolicyKey is a unique identifier for an NGF Policy.
+type PolicyKey struct {
+	// Nsname is the NamespacedName of the Policy.
+	NsName types.NamespacedName
+	// GVK is the GroupVersionKind of the Policy.
+	GVK schema.GroupVersionKind
+}
+
+const (
+	gatewayGroupKind = v1.GroupName + "/" + "Gateway"
+	hrGroupKind      = v1.GroupName + "/" + "HTTPRoute"
+)
+
+// attachPolicies attaches the graph's processed policies to the resources they target. It modifies the graph in place.
+func (g *Graph) attachPolicies() {
+	if g.Gateway == nil {
+		return
+	}
+
+	for _, policy := range g.NGFPolicies {
+		ref := policy.TargetRef
+
+		switch ref.Kind {
+		case "Gateway":
+			ancestor, attached := attachPolicyToGateway(policy, g.Gateway, g.IgnoredGateways)
+			if attached {
+				if g.Gateway.Policies == nil {
+					g.Gateway.Policies = make([]*Policy, 0, len(g.NGFPolicies))
+				}
+
+				g.Gateway.Policies = append(g.Gateway.Policies, policy)
+			}
+
+			policy.Ancestors = []PolicyAncestor{ancestor}
+		case "HTTPRoute":
+			route, exists := g.Routes[ref.Nsname]
+			if !exists {
+				return
+			}
+
+			ancestors, attached := attachPolicyToRoute(policy, route)
+
+			if attached {
+				if route.Policies == nil {
+					route.Policies = make([]*Policy, 0, len(g.NGFPolicies))
+				}
+
+				route.Policies = append(route.Policies, policy)
+			}
+
+			policy.Ancestors = ancestors
+		}
+	}
+}
+
+func attachPolicyToRoute(policy *Policy, route *Route) (ancestors []PolicyAncestor, attached bool) {
+	ancestors = make([]PolicyAncestor, 0, len(route.ParentRefs))
+
+	if len(route.ParentRefs) == 0 {
+		// this is an edge case that only happens when there are duplicate section names in the route
+		routeNsName := types.NamespacedName{Namespace: route.Source.Namespace, Name: route.Source.Name}
+		ancestors = append(ancestors, PolicyAncestor{
+			Ancestor: PolicyAncestorRef{
+				Kind:   "HTTPRoute",
+				Group:  v1.GroupName,
+				Nsname: routeNsName,
+			},
+			Conditions: []conditions.Condition{staticConds.NewPolicyTargetNotFound("TargetRef is invalid")},
+		})
+
+		return ancestors, false
+	}
+
+	for _, pr := range route.ParentRefs {
+		ancestor := PolicyAncestor{
+			Ancestor: PolicyAncestorRef{
+				Kind:        "Gateway",
```

Comment:
The more I think about this, the more I believe the ancestor should be the HTTPRoute instead of the Gateway. The policy affects the location blocks in the NGINX config which traces back to the HTTPRoute, not the Gateway. 

I think we want an entry for each HTTPRoute in the ancestor status.

Reply:
It's extremely confusing, and I've read it at least 10 times. 

Here are my notes:
- Describes the status of a policy with respect to an ancestor
- Ancestors are objects that are either the Target of a policy or above it in the hierarchy.
- Almost always, the Gateway will be the most useful object to place Policy status on. Implementations SHOULD use Gateway at the status object unless there’s a very good reason otherwise.
- Ancestor is used to distinguish which resource results in a distinct application of this policy. For example, if a policy targets a Service, it may have a distinct result per attached Gateway.
- Max number of ancestors is 16. if ancestors is full, the policy can not be implemented by any additional ancestor


I keep coming back to this bullet point:

> Ancestor is used to distinguish which resource results in a distinct application of this policy. For example, if a policy targets a Service, it may have a distinct result per attached Gateway.

For this policy, when attached to a route, there is a distinct result per Route, not per Gateway. This is especially true if we want to support multiple target refs. Let's say a policy attaches to route `foo` and route `bar`. Both routes attach to the same Gateway listener, but `foo` is invalid. What would the ancestor status be if the ancestor object is the Gateway?

Reply:
However, we also have the case where a policy targets a route that references multiple listeners on a gateway. 


Route `bar`:
- parentRefs:
  - name: gateway
     sectionName: listener-1
  - name: gateway
     sectionName: listener-2

Policy `csp`
- targetRef:
  - name: bar
     kind: HTTPRoute

Let's say `bar` attaches to the `listener-1` on the gateway, but not `listener-2`.

That means the policy will be implemented in the `listener-1` server block, but not the `listener-2` server block.

In this case, the ancestors should be the Gateway listeners:

```
- kind: Gateway
  name: gateway
  sectionName: listener-1
  conditions: 
  - Accepted/True/Accepted
- kind: Gateway
  name: gateway
  sectionName: listener-2
  conditions: 
  - Accepted/False/TargetRef failed to attach
```

Reply:
This makes me think that Client Settings shouldn't support multiple target refs. Since it can be attached at a Gateway, I don't think there's a big need for multiple target refs.

Reply:
@pleshakov 
> What if Gateway has 1000+ HTTPRoutes? Will it blow up the status?

When attaching to a Gateway, the ancestor object would be the Gateway so I don't see this as an issue.

However, I do see an issue with using HTTPRoutes as the ancestor. See my last two comments